prospero-common/src/main/java/org/wildfly/prospero/actions/ApplyPatch.java

```diff
@@ -0,0 +1,129 @@
+
+    public ApplyPatch(Path targetPath, MavenSessionManager mavenSessionManager, Console console) throws OperationException {
+        this.installDir = targetPath;
+        this.metadata = new InstallationMetadata(targetPath);
+        this.mavenSessionManager = mavenSessionManager;
+        this.console = console;
+    }
+
+    public void apply(Path patchArchivePath) throws OperationException, ProvisioningException {
+        // TODO: verify archive
+
+        try {
+            final PatchArchive patchArchive = new PatchArchive(patchArchivePath);
+            console.println(Messages.MESSAGES.installingPatch(patchArchive.getName()));
+
+            // install patch locally
+            final Patch patch = patchArchive.extract(installDir);
+
+            // update config
+            final ProvisioningConfig provisioningConfig = metadata.getProsperoConfig();
+            provisioningConfig.addChannel(new ChannelRef(null, patch.getChannelFileUrl().toString()));
+
+            // add cached repository to config if not present
+            provisioningConfig.addRepository(new RepositoryRef(PATCH_REPO_NAME, installDir.resolve(PATCHES_REPO_PATH).toUri().toURL().toString()));
+            metadata.updateProsperoConfig(provisioningConfig);
+        } catch (IOException e) {
+            e.printStackTrace();
+        } catch (MetadataException e) {
+            e.printStackTrace();
+        }
+
+        reprovisionServer();
+    }
+
+    private void reprovisionServer() throws MetadataException, ArtifactResolutionException, ProvisioningException {
+        // have to parse the config after the patch metadata was applied
+        final ProvisioningConfig prosperoConfig = metadata.getProsperoConfig();
+        final List<RemoteRepository> repositories = prosperoConfig.getRemoteRepositories();
+        final List<Channel> channels = mapToChannels(new ChannelRefUpdater(mavenSessionManager)
+                .resolveLatest(prosperoConfig.getChannels(), repositories));
+
+        final RepositorySystem system = mavenSessionManager.newRepositorySystem();
+        final DefaultRepositorySystemSession session = mavenSessionManager.newRepositorySystemSession(system);
+        final VersionResolverFactory factory = new VersionResolverFactory(system, session, repositories);
+        final ChannelSession channelSession = new ChannelSession(channels, factory);
+        ChannelMavenArtifactRepositoryManager maven = new ChannelMavenArtifactRepositoryManager(channelSession);
+        ProvisioningManager provMgr = GalleonUtils.getProvisioningManager(installDir, maven);
+
+        GalleonUtils.executeGalleon(options -> provMgr.provision(provMgr.getProvisioningConfig(), options),
+                mavenSessionManager.getProvisioningRepo().toAbsolutePath());
+
+        metadata.setChannel(maven.resolvedChannel());
+
+        metadata.writeFiles();
+    }
+
+    private List<Channel> mapToChannels(List<ChannelRef> channelRefs) throws MetadataException {
```

Comment:
Just an idea, we could create a conversion method on the ChannelRef class, e.g. `ChannelRef#toChannel()` (non-static), and then do these conversions via streams, like `channelRefs.stream().map(ChannelRef::toChannel()).collect(...)`. Or maybe have some static utility methods to convert a single instance or a whole collection. It's just that I think this method already exists in a different class. Can be a future task...

Reply:
Yes, good idea. I want to try and extract the whole section creating ProvisionManager and associated resources to utility at some point